File: backend/src/main/java/com/backend/domain/repository/service/RepositoryService.java
package com.backend.domain.repository.service;

import com.backend.domain.analysis.service.SseProgressNotifier;
import com.backend.domain.repository.dto.response.RepositoryData;
import com.backend.domain.repository.dto.response.RepositoryResponse;
import com.backend.domain.repository.dto.response.github.*;
import com.backend.domain.repository.entity.Language;
import com.backend.domain.repository.entity.Repositories;
import com.backend.domain.repository.entity.RepositoryLanguage;
import com.backend.domain.repository.repository.RepositoryJpaRepository;
import com.backend.domain.repository.repository.RepositoryLanguageRepository;
import com.backend.domain.repository.service.fetcher.GitHubDataFetcher;
import com.backend.domain.repository.service.mapper.*;
import com.backend.domain.user.entity.User;
import com.backend.domain.user.repository.UserRepository;
import com.backend.domain.user.util.JwtUtil;
import com.backend.global.exception.BusinessException;
import com.backend.global.exception.ErrorCode;
import jakarta.servlet.http.HttpServletRequest;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@Slf4j
@Service
@RequiredArgsConstructor
public class RepositoryService {

    private final UserRepository userRepository;
    private final GitHubDataFetcher gitHubDataFetcher;
    private final RepositoriesMapper repositoriesMapper;
    private final RepositoryInfoMapper repositoryInfoMapper;
    private final CommitInfoMapper commitInfoMapper;
    private final ReadmeInfoMapper readmeInfoMapper;
    private final SecurityInfoMapper securityInfoMapper;
    private final TestInfoMapper testInfoMapper;
    private final CicdInfoMapper cicdInfoMapper;
    private final IssueInfoMapper issueInfoMapper;
    private final PullRequestInfoMapper pullRequestInfoMapper;
    private final RepositoryJpaRepository repositoryJpaRepository;
    private final SseProgressNotifier sseProgressNotifier;
    private final RepositoryLanguageRepository repositoryLanguageRepository;
    private final JwtUtil jwtUtil;

    @Transactional
    public RepositoryData fetchAndSaveRepository(String owner, String repo, Long userId) {
        try {
            return fetchCompleteRepositoryData(owner, repo, userId);
        } catch (BusinessException e) {
            throw e;
        } catch (Exception e) {
            log.error("❌ Repository 수집 중 오류 발생", e);
            throw new BusinessException(ErrorCode.INTERNAL_ERROR);
        }
    }

    // Github 저장소 데이터를 수집하고 DB에 저장 및 RepositoryData로 매핑
    @Transactional
    public RepositoryData fetchCompleteRepositoryData(String owner, String repo, Long userId) {
        RepositoryData data = new RepositoryData();

        try {
            // 1. 기본 정보 수집 및 매핑 + Repositories 테이블 저장
            safeSendSse(userId, "status", "GitHub 연결 중");
            RepoResponse repoInfo = gitHubDataFetcher.fetchRepositoryInfo(owner, repo);
            validateRepositorySize(repoInfo.size());
            repositoryInfoMapper.mapBasicInfo(data, repoInfo);

            // 2. 커밋 데이터 수집 및 매핑
            safeSendSse(userId, "status", "커밋 히스토리 분석");
            ZonedDateTime ninetyDaysAgoUtc = ZonedDateTime.now(ZoneOffset.UTC).minus(90, ChronoUnit.DAYS);
            String sinceParam = ninetyDaysAgoUtc.format(DateTimeFormatter.ISO_INSTANT);
            List<CommitResponse> commits = gitHubDataFetcher.fetchCommitInfo(owner, repo, sinceParam);
            commitInfoMapper.mapCommitInfo(data, commits);

            // 3. README 데이터 수집 및 매핑
            safeSendSse(userId, "status", "문서화 품질 분석");
            String readme = gitHubDataFetcher.fetchReadmeContent(owner, repo).orElse("");
            readmeInfoMapper.mapReadmeInfo(data, readme);

            // 4. 보안 관리 데이터 수집 및 매핑
            safeSendSse(userId, "status", "보안 구성 분석");
            TreeResponse tree = gitHubDataFetcher.fetchRepositoryTreeInfo(owner, repo, repoInfo.defaultBranch()).orElse(null);
            securityInfoMapper.mapSecurityInfo(data, tree);

            // 5. 테스트 데이터 수집 및 매핑
            safeSendSse(userId, "status", "테스트 구성 분석");
            testInfoMapper.mapTestInfo(data, tree);

            // 6. CI/CD 데이터 수집 및 매핑
            safeSendSse(userId, "status", "CI/CD 설정 분석");
            cicdInfoMapper.mapCicdInfo(data, tree);

            // 7. 커뮤니티 활성도 데이터 수집 및 매핑
            safeSendSse(userId, "status", "커뮤니티 활동 분석");
            List<IssueResponse> issues = gitHubDataFetcher.fetchIssueInfo(owner, repo);
            issueInfoMapper.mapIssueInfo(data, issues);
            List<PullRequestResponse> prs = gitHubDataFetcher.fetchPullRequestInfo(owner, repo);
            pullRequestInfoMapper.mapPullRequestInfo(data, prs);

            // Entity 저장 로직
            saveOrUpdateRepository(repoInfo, owner, repo, userId);

            return data;
        } catch (BusinessException e) {
            safeSendSse(userId, "error", "❌ " + e.getErrorCode().getMessage());
            throw e;

        } catch (Exception e) {
            safeSendSse(userId, "error", "❌ Repository 데이터 수집 실패: " + e.getMessage());
            throw e;
        }
    }

    // SSE 전송 헬퍼 메서드
    private void safeSendSse(Long userId, String event, String message) {
        try {
            sseProgressNotifier.notify(userId, event, message);
        } catch (Exception e) {
            log.warn("SSE 전송 실패 (분석은 계속): userId={}, event={}, error={}",
                    userId, event, e.getMessage());
        }
    }

    /* Repository Entity를 DB에 저장하거나 기존 데이터 업데이트
    * 같은 htmlUrl + userId 조합이 존재하면 업데이트, 없으면 신규 데이터 저장 */
    private void saveOrUpdateRepository(RepoResponse repoInfo, String owner, String repo, Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new BusinessException(ErrorCode.USER_NOT_FOUND));

        Map<String, Integer> languagesData = gitHubDataFetcher.fetchLanguages(owner, repo);

        repositoryJpaRepository.findByHtmlUrlAndUserId(repoInfo.htmlUrl(), userId)
                .ifPresentOrElse(existing -> {
                    existing.updateFrom(repoInfo);
                    existing.updateLanguagesFrom(languagesData);
                },
                () -> {
                    Repositories newRepo = repositoriesMapper.toEntity(repoInfo, user);
                    newRepo.updateLanguagesFrom(languagesData);
                    repositoryJpaRepository.save(newRepo);
                });
    }

    // 특정 사용자의 모든 Repository 조회
    public List<Repositories> findRepositoryByUser(Long userId) {
        return repositoryJpaRepository.findByUserId(userId);
    }

    // Repository ID로 단건 조회
    public Optional<Repositories> findById(Long repositoriesId) {
        return repositoryJpaRepository.findById(repositoriesId);
    }

    // 저장소 크기 검증
    private void validateRepositorySize(Integer sizeInKb) {
        if (sizeInKb == null) {
            return;
        }

        final int MAX_SIZE_KB = 1_000_000;

        if (sizeInKb > MAX_SIZE_KB) {
            log.warn("저장소 크기 초과: {}KB (제한: {}KB)", sizeInKb, MAX_SIZE_KB);
            throw new BusinessException(ErrorCode.GITHUB_REPO_TOO_LARGE);
        }
    }

    // repository 사용 언어 반환
    public List<Language> getLanguageByRepositoriesId(Long repositoriesId) {
        return repositoryLanguageRepository.findByRepositories_Id(repositoriesId)
                .stream()
                .map(RepositoryLanguage::getLanguage)
                .toList();
    }

    // 사용자 분석 결과 조회
    @Transactional(readOnly = true)
    public List<RepositoryResponse> getUserRepositories(HttpServletRequest request) {
        Long userId = jwtUtil.getUserId(request);
        if (userId == null) {
            throw new BusinessException(ErrorCode.UNAUTHORIZED);
        }

        return findRepositoryByUser(userId)
                .stream()
                .map(RepositoryResponse::new)
                .toList();
    }
}

File: backend/src/main/java/com/backend/domain/repository/service/RepositoryService.kt
package com.backend.domain.repository.service

import com.backend.domain.analysis.service.SseProgressNotifier
import com.backend.domain.repository.dto.response.RepositoryData
import com.backend.domain.repository.dto.response.RepositoryResponse
import com.backend.domain.repository.dto.response.github.RepoResponse
import com.backend.domain.repository.entity.Language
import com.backend.domain.repository.entity.Repositories
import com.backend.domain.repository.entity.RepositoryLanguage
import com.backend.domain.repository.repository.RepositoryJpaRepository
import com.backend.domain.repository.repository.RepositoryLanguageRepository
import com.backend.domain.repository.service.fetcher.GitHubDataFetcher
import com.backend.domain.repository.service.mapper.*
import com.backend.domain.user.repository.UserRepository
import com.backend.domain.user.util.JwtUtil
import com.backend.global.exception.BusinessException
import com.backend.global.exception.ErrorCode
import jakarta.servlet.http.HttpServletRequest
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.ZoneOffset
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit

@Service
class RepositoryService(
    private val userRepository: UserRepository,
    private val gitHubDataFetcher: GitHubDataFetcher,
    private val repositoriesMapper: RepositoriesMapper,
    private val repositoryInfoMapper: RepositoryInfoMapper,
    private val commitInfoMapper: CommitInfoMapper,
    private val readmeInfoMapper: ReadmeInfoMapper,
    private val securityInfoMapper: SecurityInfoMapper,
    private val testInfoMapper: TestInfoMapper,
    private val cicdInfoMapper: CicdInfoMapper,
    private val issueInfoMapper: IssueInfoMapper,
    private val pullRequestInfoMapper: PullRequestInfoMapper,
    private val repositoryJpaRepository: RepositoryJpaRepository,
    private val sseProgressNotifier: SseProgressNotifier,
    private val repositoryLanguageRepository: RepositoryLanguageRepository,
    private val jwtUtil: JwtUtil
) {
    private val log = LoggerFactory.getLogger(RepositoryService::class.java)

    @Transactional
    fun fetchAndSaveRepository(owner: String, repo: String, userId: Long): RepositoryData {
        return try {
            fetchCompleteRepositoryData(owner, repo, userId)
        } catch (e: BusinessException) {
            throw e
        } catch (e: Exception) {
            log.error("❌ Repository 수집 중 오류 발생", e)
            throw BusinessException(ErrorCode.INTERNAL_ERROR)
        }
    }

    @Transactional
    fun fetchCompleteRepositoryData(owner: String, repo: String, userId: Long): RepositoryData {
        val data = RepositoryData()

        try {
            // 1. 기본 정보 수집
            safeSendSse(userId, "status", "GitHub 연결 중")

            val repoInfo = gitHubDataFetcher.fetchRepositoryInfo(owner, repo)
                ?: throw BusinessException(ErrorCode.GITHUB_REPO_NOT_FOUND)

            validateRepositorySize(repoInfo.size)
            repositoryInfoMapper.mapBasicInfo(data, repoInfo)

            // 2. 커밋
            safeSendSse(userId, "status", "커밋 히스토리 분석")
            val since = ZonedDateTime.now(ZoneOffset.UTC)
                .minus(90, ChronoUnit.DAYS)
                .format(DateTimeFormatter.ISO_INSTANT)

            val commits = gitHubDataFetcher.fetchCommitInfo(owner, repo, since)
            commitInfoMapper.mapCommitInfo(data, commits)

            // 3. README
            safeSendSse(userId, "status", "문서화 품질 분석")
            val readme = gitHubDataFetcher.fetchReadmeContent(owner, repo) ?: ""
            readmeInfoMapper.mapReadmeInfo(data, readme)

            // 4. 트리 정보
            safeSendSse(userId, "status", "보안 구성 분석")
            val tree = gitHubDataFetcher.fetchRepositoryTreeInfo(
                owner,
                repo,
                repoInfo.defaultBranch ?: "main"
            )

            securityInfoMapper.mapSecurityInfo(data, tree)

            // 5. 테스트
            safeSendSse(userId, "status", "테스트 구성 분석")
            testInfoMapper.mapTestInfo(data, tree)

            // 6. CI/CD
            safeSendSse(userId, "status", "CI/CD 설정 분석")
            cicdInfoMapper.mapCicdInfo(data, tree)

            // 7. Issue/PR
            safeSendSse(userId, "status", "커뮤니티 활동 분석")

            val issues = gitHubDataFetcher.fetchIssueInfo(owner, repo)
            issueInfoMapper.mapIssueInfo(data, issues)

            val prs = gitHubDataFetcher.fetchPullRequestInfo(owner, repo)
            pullRequestInfoMapper.mapPullRequestInfo(data, prs)

            // 8. Entity 저장
            saveOrUpdateRepository(repoInfo, owner, repo, userId)

            return data

        } catch (e: BusinessException) {
            safeSendSse(userId, "error", "❌ ${e.errorCode.message}")
            throw e
        } catch (e: Exception) {
            safeSendSse(userId, "error", "❌ Repository 데이터 수집 실패: ${e.message}")
            throw e
        }
    }

    private fun safeSendSse(userId: Long, event: String, message: String) {
        try {
            sseProgressNotifier.notify(userId, event, message)
        } catch (e: Exception) {
            log.warn("SSE 전송 실패: userId={}, event={}, error={}", userId, event, e.message)
        }
    }

    private fun saveOrUpdateRepository(
        repoInfo: RepoResponse,
        owner: String,
        repo: String,
        userId: Long
    ) {
        val user = userRepository.findById(userId)
            .orElseThrow { BusinessException(ErrorCode.USER_NOT_FOUND) }

        val languages = gitHubDataFetcher.fetchLanguages(owner, repo)

        val url = repoInfo.htmlUrl
            ?: throw BusinessException(ErrorCode.INVALID_GITHUB_URL)

        val existing = repositoryJpaRepository.findIncludingDeleted(url, userId)

        if (existing != null) {

            if (existing.deleted) {
                existing.deleted = false
                repositoryJpaRepository.save(existing)
            }

            existing.updateFrom(repoInfo)
            existing.updateLanguagesFrom(languages)

            return
        }

        val newRepo = repositoriesMapper.toEntity(repoInfo, user)
        newRepo.updateLanguagesFrom(languages)

        repositoryJpaRepository.save(newRepo)
    }

    fun findRepositoryByUser(userId: Long): List<Repositories> =
        repositoryJpaRepository.findByUserId(userId)

    fun findById(repositoriesId: Long) =
        repositoryJpaRepository.findById(repositoriesId)

    private fun validateRepositorySize(sizeInKb: Int?) {
        val max = 1_000_000
        if (sizeInKb != null && sizeInKb > max) {
            log.warn("저장소 크기 초과: {}KB (제한 {}KB)", sizeInKb, max)
            throw BusinessException(ErrorCode.GITHUB_REPO_TOO_LARGE)
        }
    }

    fun getLanguageByRepositoriesId(repositoriesId: Long): List<Language> =
        repositoryLanguageRepository.findByRepositories_Id(repositoriesId)
            .map(RepositoryLanguage::language)

    @Transactional(readOnly = true)
    fun getUserRepositories(request: HttpServletRequest): List<RepositoryResponse> {
        val userId = jwtUtil.getUserId(request)
            ?: throw BusinessException(ErrorCode.UNAUTHORIZED)

        return findRepositoryByUser(userId)
            .map { RepositoryResponse.from(it) }
    }
}
